### ISSUE TITLE
Can't startup backend locally a lot of errors

### ISSUE THREAD

**Issue by vadim9999** [atherdon/dash-back#27]

**Comment by vadim9999** [comment]
@kolserdav I am on win 10 now, and I can't startup backend. First of all, I installed cross-env then nodemon. 
![image](https://user-images.githubusercontent.com/35640573/107393170-88de1480-6b0b-11eb-8707-295ad2b6e0b2.png)
And tried to run project `npm run dev` without success. 
I started to build it and got errors
![image](https://user-images.githubusercontent.com/35640573/107393338-bf1b9400-6b0b-11eb-9239-5a1937000816.png)

**Comment by vadim9999** [comment]
@kolserdav Am I right understood that I can get types from your backend?

**Comment by kolserdav** [comment]
> @kolserdav I am on win 10 now, and I can't startup backend. First of all, I installed cross-env then nodemon.
> ![image](https://user-images.githubusercontent.com/35640573/107393170-88de1480-6b0b-11eb-8707-295ad2b6e0b2.png)
> And tried to run project `npm run dev` without success.
> I started to build it and got errors
> ![image](https://user-images.githubusercontent.com/35640573/107393338-bf1b9400-6b0b-11eb-9239-5a1937000816.png)

What is your version of the node?

**Comment by kolserdav** [comment]
> @kolserdav Am I right understood that I can get types from your backend?

Yes, all graphql types define in the file `src/types/graphql/index.d.ts`

**Comment by vadim9999** [comment]
@kolserdav my version is 14.15.1

**Comment by kolserdav** [comment]
Before `dev` or `start` run you `ts` or `build`?

**Comment by kolserdav** [comment]
> Before `dev` or `start` run you `ts` or `build`?

Oh i see that 'yes
'

**Comment by vadim9999** [comment]
![Uploading image.png…]()

**Comment by vadim9999** [comment]
I don't have the file dash.db

**Comment by kolserdav** [comment]
> I don't have the file dash.db. Where can I find it?

Need run `migrate:dev` [Migration on development](https://github.com/atherdon/dash-back#migration-on-development)

**Comment by kolserdav** [comment]
> > @kolserdav I am on win 10 now, and I can't startup backend. First of all, I installed cross-env then nodemon.
> > ![image](https://user-images.githubusercontent.com/35640573/107393170-88de1480-6b0b-11eb-8707-295ad2b6e0b2.png)
> > And tried to run project `npm run dev` without success.
> > I started to build it and got errors
> > ![image](https://user-images.githubusercontent.com/35640573/107393338-bf1b9400-6b0b-11eb-9239-5a1937000816.png)
> 
> What is your version of the node?

Will be fixed in the next version. Wait a bit.

**Comment by vadim9999** [comment]
![image](https://user-images.githubusercontent.com/35640573/107406516-8f738880-6b19-11eb-8671-2e2c0a15b0c6.png)

**Comment by kolserdav** [comment]
Now I need to try to run in Windows myself

**Comment by atherdon** [comment]
or we can move forward [preview/deploys discussion](https://github.com/atherdon/dash-back/discussions/21). Because I still wish to be able to have a quick setup where I can check/test updates without doing it on my local machine

**Comment by kolserdav** [comment]
> 
> 
> ![image](https://user-images.githubusercontent.com/35640573/107406516-8f738880-6b19-11eb-8671-2e2c0a15b0c6.png)

For Windows, in the path you need to use `\` instead of `/`. But, as I see further, there is a prism engine error on Windows.

**Comment by kolserdav** [comment]
Now I started the server at https://dash-back.uyem.ru/graphql, for the front-end it should be enough for now, I will deal with Windows myself

**Comment by vadim9999** [comment]
Ok, thank you

**Comment by vadim9999** [comment]
am I right understood, that it's returns empty array? 
```
query {
    getManyAuthor {
        id
    }
}
```
How can I see documentation of queries? I mean .../graphiql

**Comment by kolserdav** [comment]
[API.md](https://github.com/atherdon/dash-back/blob/main/docs/API.md)

**Comment by kolserdav** [comment]
Database is empty, before `getMany` need create new with `postOne`

**Comment by kolserdav** [comment]
Some links in 1 do not work correctly, for now find the files you need manually

**Comment by vadim9999** [comment]
How can I do request on frontend side? Should I use appolo client?

**Comment by kolserdav** [comment]
> 
> 
> How can I do request on frontend side? Should I use appolo client?

As you prefer. But apollo server is a good solution.

**Comment by kolserdav** [comment]
Wrote in the repository prisma https://github.com/prisma/prisma/issues/3313

**Comment by vadim9999** [comment]
@kolserdav  Ok I tried to run it on Linux Mint 19.3 it doesn't work 

![image](https://user-images.githubusercontent.com/35640573/107525036-19296180-6bbf-11eb-8bea-105e629fff9e.png)

**Comment by vadim9999** [comment]
Build not working but 
`npm run migrate:dev` and `npm run dev` worked for me

**Comment by kolserdav** [comment]
> @kolserdav Ok I tried to run it on Linux Mint 19.3 it doesn't work
> 
> ![image](https://user-images.githubusercontent.com/35640573/107525036-19296180-6bbf-11eb-8bea-105e629fff9e.png)

```typescript
// Auth middleware
//app.use(server.graphqlPath, middlewares.auth); // comment this line
```
Yes, this is a bug of an unfinished middleware. You can comment out his call in [src/index.ts](https://github.com/atherdon/dash-back/blob/main/src/index.ts)

**Comment by kolserdav** [comment]
> Build not working but
> `npm run migrate:dev` and `npm run dev` worked for me

In version `3.3.0` this bug was fixed.

**Comment by vadim9999** [comment]
@kolserdav  I connected Apollo client but have errors with CORS. Could you enable cors in dev mode?

**Comment by vadim9999** [comment]
![image](https://user-images.githubusercontent.com/35640573/107673119-5a447300-6c9e-11eb-9646-8342e701b5be.png)

**Comment by vadim9999** [comment]
I have just added this code temporary for developing and it helped me

![image](https://user-images.githubusercontent.com/35640573/107673673-e0f95000-6c9e-11eb-9fc3-2b18f20a8432.png)

**Comment by kolserdav** [comment]
in file `src/index.ts`
```typescript
server.applyMiddleware({
  app,
  cors: { credentials: true, origin: lib.isDev() ? '*' : CORS_ORIGIN },
});
```

**Comment by kolserdav** [comment]
thans, i fix it in version `3.3.1`
```typescript
app.use(
  cors({
    credentials: true,
    origin: lib.isDev() ? '*' : CORS_ORIGIN,
  })
);
```

**Comment by vadim9999** [comment]
@kolserdav do you have ideas how we can remove Maybe From generated types?
![image](https://user-images.githubusercontent.com/35640573/107780755-ccc15b80-6d4f-11eb-9e11-b4d147009ba3.png)

**Comment by vadim9999** [comment]
@kolserdav never mind I have found other solution. I am using typescript-react-apollo and using generated hooks 
![image](https://user-images.githubusercontent.com/35640573/107797083-3434d680-6d63-11eb-9963-f7fb9cdbaaa4.png)

**Comment by kolserdav** [comment]
The types are generated using `@codegen/cli`. They only help to work with data when preparing a query and processing the result, but they do not contain programming logic.

**Comment by kolserdav** [comment]
Please show your `[model].graphql` file

**Comment by kolserdav** [comment]
Thank. Your additions seem like a good solution, I will consider including this in future versions.

**Comment by kolserdav** [comment]
> Thank. Your additions seem like a good solution, I will consider including this in future versions.

I have not found the benefits of embedding `typescript-react-apollo` in the backend. Since this does nothing on the backend, but it forces me to duplicate executable requests from `postman.json`. Therefore, I propose to do the generation of mutations separately on the frontend.

**Comment by atherdon** [comment]
I like the idea to separate stuff into simple chunks that only doing one thing.

**Comment by vadim9999** [comment]
> > Thank. Your additions seem like a good solution, I will consider including this in future versions.
> 
> I have not found the benefits of embedding `typescript-react-apollo` in the backend. Since this does nothing on the backend, but it forces me to duplicate executable requests from `postman.json`. Therefore, I propose to do the generation of mutations separately on the frontend.

@kolserdav what do you mean about removing typescript-react-apollo. I am using this library on frontend side. if you delete this from backend I will not be able to use react-apollo hooks?
Could you add please to Readme instruction how to run this backend without explanation. I mean just commands.
Also add query for getting all this counters
![image](https://user-images.githubusercontent.com/35640573/108520477-64382880-72d3-11eb-8e57-7e84e6569db0.png)

**Comment by vadim9999** [comment]
@kolserdav what format do you use for storing date? it should be in ISO format stored in DB 
![image](https://user-images.githubusercontent.com/35640573/108523922-263d0380-72d7-11eb-944d-55914a9ca8ea.png)
Could you add validation for this data type?

**Comment by kolserdav** [comment]
`created` and `updated` are not available for api, the server records them on its own.  when recording and updating.  they are not related to materials, they are only timestamps of operations with the base.  and the timestamps of materials in their original form only had a month and a day, so they could not be converted to `DateTime` type

**Comment by vadim9999** [comment]
yeah I understand but I think it should be validation for date type

**Comment by kolserdav** [comment]
>what do you mean about removing typescript-react-apollo. I am using this library on frontend side. if you delete this from backend I will not be able to use react-apollo hooks?
>
`typescript-react-apollo` was not installed on the backend, to use mutation hooks you need to enable `useHooks` in `codegen.yml` and write a schema.graphql `document` with named queries.
Example name.graphql file:
```
query getManyBrand {
  id
  ...
}
mutation postOneBrand($data: PostOneBrandParams!) {
  postOneBrand(data: $data) {
        id
         ...
    }
}
```

**Comment by vadim9999** [comment]
![image](https://user-images.githubusercontent.com/35640573/108531323-3c4ec200-72df-11eb-9a70-717269ef6d8a.png)

I installed this config a week ago on frontend side and query hooks works good

**Comment by kolserdav** [comment]
> @kolserdav what format do you use for storing date? it should be in ISO format stored in DB
> ![image](https://user-images.githubusercontent.com/35640573/108523922-263d0380-72d7-11eb-944d-55914a9ca8ea.png)
> Could you add validation for this data type?

Graphql workin through http and it not have support for type Date, only serialized to ISO string for this case

**Comment by kolserdav** [comment]
but you can trust this, everything is checked inside and will definitely return in iso format

**Comment by kolserdav** [comment]
probably you were confused by this. 
```
    created: result.created?.toISOString() || '',
    updated: result.updated?.toISOString() || '',
```
 this is an operation to bypass type errors, since the now () function is set by default for these fields in the database, then prisma generates their types as optional, but in fact they cannot be empty in any way)

**Comment by vadim9999** [comment]
@kolserdav can't run this backend
I have done 
npm run migrate:dev
npm run build

![image](https://user-images.githubusercontent.com/35640573/108866005-76c89f80-75fc-11eb-9d6b-bdfbb227a367.png)

**Comment by vadim9999** [comment]
Also I stiil can't undertand how to run your project. Did you wrote in Readme small snippet how to startup it?

**Comment by kolserdav** [comment]
File expandable2 will updated without me( i can't help you.

**Comment by vadim9999** [comment]
@atherdon Do you mean Arthur broke this backend?

**Comment by kolserdav** [comment]
Types not checked before commit, and that is result.

**Comment by kolserdav** [comment]
You can comment fillExpandable in src/bin/index.ts

**Comment by vadim9999** [comment]
@kolserdav Could you fix this please?

**Comment by kolserdav** [comment]
Yes tomorrow, and now simple do coment:
```
  await fillExpandable();
```
In src/bin/index.ts
